README.md
# Installation


## Preconditions

1. You have configured your Apache as follows:

```
<VirtualHost IP:443> 
  ...
  # this is under virtualhost :443 section
  # your webserver public and private cert
  SSLCertificateFile    /etc/ssl/certs/www.domain.ee.crt
  SSLCertificateKeyFile /etc/ssl/private/www.domain.ee.key
  # This makes certain, that its valid smartcard and is allowed (google for ID kaart + apache server)
  SSLCACertificateFile /etc/ssl/certs/id.crt
  ...
</VirtualHost>
```

If you want that before opening Wiki user certificate is asked and autolog is done, add also these lines to your Virtualhost config or .htaccess file in Wiki root:

```
# request client cert
SSLVerifyClient require
SSLVerifyDepth 2
SSLOptions +StdEnvVars +ExportCertData
```

2. Using .htaccess must be allowed. If it is not allowed, you have to copy inc/auth/smartcard/.htaccess content to virtualhost directory section for inc/auth/smartcard:

```
<VirtualHost IP:443>
  ...
  # Usally using Directory instead of Location is recommended, 
  # but this is not for security, but for asking certificate
  <Location "/inc/auth/smartcard">
    SSLVerifyClient Optional
    SSLVerifyDepth 2
    SSLOptions +StdEnvVars +ExportCertData
  </Location>
...
</VirtualHost>
```



3. If your value against what your would like to check is not in certificate.subject.serialNumber, you will have to change `inc/auth/smartcard.php` line 85.

## Steps

- cd your_dokuwiki_folder
- svn export https://dokuwiki-smartcard-authentication.googlecode.com/svn/trunk/dokuwiki/ . --force
- Configure authentication module (See: "Configure parameters")
- Make Smartcard autentication available for user
  - Add to your wiki what would reference to https://YOUR_DOMAIN/DOKUWIKI_PATH/inc/auth/smartcard
  - Modify in YOUR_DOKUWIKI_INSTALLATION/inc/lang/YOUR_CHOSEN_LANGUAGE/login.txt and add line: `To log in with smart card, click: [[inc/auth/smartcard|Autenticate me with Smartcard]]`
- Clear your Dokuwiki cache 



## Allowing Smartcard authentication

You have to do:
- Mark user serial (or some other parameter for check) as one of his groups.

This auth module will authenticated user based on the serial found on smartcard with steps:

- If ?u=smartcard&p=smartcard is called script check that `certificate.subject.serialNumber` exists (from `$_SERVER['SSL_CLIENT_S_DN'`)
  - if it exists: user, who has this value as his group, is allowed in
  - if it does not exist: user is not allowed in
- If auth is called with other parameters and `$conf['smartcard']['allow_without_smartcard']` is checked
  - if it is `true`: auth is delegated to auth modules defined in `$conf['smartcard']['use_authtypes']`
  - else login with username and password will not be possible.


## Configuration

Must be defined in file: YOUR_WIKI_INSTALLATION/conf/local.php.

Config parameters explanation:

```php
# set for docuwiki that smartcard auth module is used
$conf['authtype']   = 'smartcard';

# if this is set to true, without smartcard login is not allowed
$conf['auth']['smartcard']['allow_without_smartcard'] = 'true';

# auth modules to use. separated with comma (",")
$conf['auth']['smartcard']['use_authtypes'] = 'plain,mysql';

# log debug info to file
$conf['auth']['smartcard']['log_to_file'] = 'true';
```


# Known problems

- When using Fckglite WYSIWYG editor, first login fails - because fckglite resets session - fix info: http://code.google.com/p/dokuwiki-smartcard-authentication/issues/detail?id=1


# Author

Margus Pärt (mxrguspxrt)


CHANGES.txt
  * 2011 07 20 - Added canDo['logout']=>true for 2011-05-25, so logging out would work
  * 2011 02 15 - SSL_CLIENT_S_DN instead of SSL_CLIENT_CERT is used (due
  * headers can't be multiline and behind proxy SSL_CLIENT_CERT is not
  * available in correct format), added few language files for en, et, uk
  * and ru.
  * 2010 05 22 - Added info message if Smartcard cert was not found
  * 2010 05 22 - Added session_start() to inc/auth/smartcard/index.php to
  * solve auth fail when session.auto_start=0
  * 2010 05 18 - Fixed that some plugin forced re-auth, before session and
  * cookies were set, auth failed.
  * 2010 04 26 - initial release


dokuwiki/conf/auth_smartcard.php
<?php


$conf['authtype']   = 'smartcard';


// this creates empty array for smartcard conf plugin
$conf['smartcard']                            = array();
// if this is set to true, without smartcard login is not allowed
$conf['smartcard']['allow_without_smartcard'] = true;
// if user tries to log in with smartcard, all those auth modules are used
$conf['smartcard']['use_authtypes']           = array('plain');
// if we want to log all authentication try info to http://www.mywikidomain.ee/log/auth_smartcard.log then we set this value to true (hide this page using ACL from your users and enable to admins only)
$conf['smartcard']['log_file']                = true;




// 20100602 this is fix for fckglite, see http://code.google.com/p/dokuwiki-smartcard-authentication/issues/detail?id=1
if(!isset($conf['csrf']) || !$conf['csrf']){
    $conf['csrf'] = md5(rand());
} 
  // /this is fix for fckglite
  

dokuwiki/inc/auth/smartcard/index.php
<?php

/*
* @version    2011 02 15
* @author     mxrgus.pxrt <margus@tione.eu>
**/

// include dokuwiki ness things (sets session.id also)
if(!defined('DOKU_INC')) define('DOKU_INC',dirname(__FILE__).'/../../../');
require_once(DOKU_INC.'inc/init.php');
require_once(DOKU_INC.'inc/common.php');
require_once(DOKU_INC.'inc/events.php');
require_once(DOKU_INC.'inc/pageutils.php');

// if smartcard auth was success, write SSL_CLIENT_CERT to session (it will be used from auth_smartcard later on)
if(isset($_SERVER['SSL_CLIENT_CERT'])){
  session_start();  

  $_SESSION['SSL_CLIENT_S_DN'] =$_SERVER['SSL_CLIENT_S_DN'];
  $_SESSION['SSL_CLIENT_CERT'] =$_SERVER['SSL_CLIENT_CERT'];
  
  session_write_close();
}

// redirect back to dokuwiki main url
$gotowikiurl = "../../../doku.php?u=smartcard&p=smartcard&id=" . getID();
header("Location: $gotowikiurl");



dokuwiki/inc/auth/smartcard.class.php
<?php
/**
  * Smartcard authentication backend
  *
  * This class itself does not provide any user managment functionality, only searches them from other backends, and forwards user managment operations to them (tested only with plain backend)
  *
  * Quick overview.
  *  
  * Methods:
  * - ____call (calls method on all auth modules marked in $conf['auth']['smartcard']['use_authtypes']
  * - __call (this is magical method called always, if on this class haven't defined this method yet, methodname and args are passed to ____calls but only one value is returned. This is why I did not extend auth_basic.)
  * - checkPass (trys to autenticate with smartcard or if allowed, with password - search is done againts all backends)
  * 
  *
  * @version 2010 05 22
  * @author     mxrgus.pxrt <margus@tione.eu>
  */


// 20100602 this is fix for fckglite, see http://code.google.com/p/dokuwiki-smartcard-authentication/issues/detail?id=1
if(!isset($conf['csrf']) || !$conf['csrf']){
  $conf['csrf'] = md5(rand());
} 
    




define('DOKU_AUTH', dirname(__FILE__));

class auth_smartcard {

  var $success      = true;
  var $cando        = array (
    'addUser'     => true, // can Users be created?
    'delUser'     => true, // can Users be deleted?
    'modLogin'    => true, // can login names be changed?
    'modPass'     => true, // can passwords be changed?
    'modName'     => true, // can real names be changed?
    'modMail'     => true, // can emails be changed?
    'modGroups'   => true, // can groups be changed?
    'getUsers'    => true, // can a (filtered) list of users be retrieved?
    'getUserCount'=> true, // can the number of users be retrieved?
    'getGroups'   => true, // can a list of available groups be retrieved?
    'external'    => false, // does the module do external auth checking?
    'logoff'      => false, // has the module some special logoff method?
    'UserMod'     => true,
    'logout'      => true
  );



  function canDo($what){
    return $this->cando[$what];
  }



  /**
   * Constructor.
   *
   *
   */
  function __construct() {
  }



  /*
  * checkPass
  *
  *
  * This method can cause problems if auth_login function changes a lot (probleby will not change so much), if from auth_login $auth->checkPass 
  * (This method is called, after getting value true back, it makes auth true, it uses input args previosly
  * used on this method and creates session and sets user cookies - problem is that Smartcard user
  * havent given them so we'll write them over here and auth can to is trick.
  *
  * I did not use trustExternal because when writing session and cookie info myself - right click download on file would disconnect session on IE7
  * Also this way seemed more cleaner.
  *
  * @param &$user - as smartcard user useses u=smartcard, we will get value from auth_login replace it and auth_login does it tricks as it should
  * @param &$password - as smartcard user useses p=smartcard, we will get value from auth_login replace it and auth_login does it tricks as it should
  */
  function checkPass(&$username, &$password){
    global $conf;
 
    session_start();
    // ALREADY LOGGED IN - we mess up the password, so we dont really know what it is, and need to get it from session
    if(isset($_SESSION['smartcard_userdata']) && $_SESSION['smartcard_userdata']['username']==$username && md5($_SESSION['smartcard_userdata']['pass'])==$password){
      return true;
    }

    // USERNAME AND PASSWORD LOGIN
    // if password login - auth with username and password
    if($username && $password && $username!='smartcard'){
      // find user by username and password
      $userdata  = $this->findUserByUsernameAndPassword($username, $password);
      // if userdata not found
      if(!$userdata){
        return false;
      }
    }

    // SMARTCARD LOGIN
    // if smartcard login - auth with $_SESSION['SSL_CLIENT_S_DN']
    if($username=='smartcard'){
      // if client cert does not exist
      if(!(isset($_SESSION['SSL_CLIENT_S_DN']) && $_SESSION['SSL_CLIENT_S_DN'])){
        msg('Smartcard was not found. Please check that it is connected and drivers are installed.', -1);
        return false;
      }
      // get serial from dn
      preg_match('/(?<=serialNumber=)\d+/', $_SESSION['SSL_CLIENT_S_DN'], $serial_in_array);
      $serial = $serial_in_array[0];
      // if serial not found
      if(!$serial){
        msg('Did not find valid serial.', -1);
        return false;
      }
      // find user by serial
      $userdata = $this->findUserBySerial($serial);
      // if userdata not found 
      if(!$userdata){ 
        msg('Did not find user with serial: '.$serial, -1); 
        return false;
      }
    }

    // SEE THAT WE GOT SMTH AND LOG HIM IN
    if($userdata){
      // write over gotten args (notice, function was defined: checkPass(&$username, &$password))
      $username           = $userdata['username'];
      $password           = md5($userdata['pass']);
      // set $_SESSION['smartcard_userdata'] because else auth will fail later on because unvalid pw later
      $_SESSION['smartcard_userdata'] = $userdata;
      session_write_close();
      return true;
    }

    // COULD NOT DO LOGIN
    unset($_SESSION['smartcard_userdata']);
    session_write_close();
    return false;
  }



  /**
  * Finds user by username and password
  *
  */
  function findUserByUsernameAndPassword($username, $password){
    // clean
    $username = preg_replace('/[^\w\d\.-_]/', '', $username);
    $password = preg_replace('/[^\w\d\.-_]/', '', $password);

    // it could cause problem that user is defined in two backends (ie plain and mysql) - mysql and plain pw different, auth is succeeds againts mysql, but user data is loaded from plain.
    if($this->__call('checkPass', array($username, $password))){
      // find userdata and return it
      $this->__log(array(__METHOD__, "Found.", $userdata));
      $userdata = $this->__call('getUserData', array($username));
      $userdata['username'] = $username;
      return $userdata;
    }else{
      // if not found
      $this->__log(array(__METHOD__, "Did not find.", $username));
      return false;
    }  
  }



  /**
  * Finds user by serial
  *
  */
  function findUserBySerial($serial){
    // Log and return false if unvalid format
    if(preg_match('/[^\w\d\.]/', $serial) || !$serial){
      $this->__log(array(__METHOD__, "$serial is not valid serial"));
      return false;
    }
    
    // find users -> resolves to __call -> ____call
    $users    = $this->__call('retrieveUsers', array(0, 2000, array('grps'=>$serial))); // receives users only from first place from where it can find them (ideally when serial ok, only 1)

    // if user count 1
    if(count($users)==1){  
      // create username value for user    
      foreach($users as $key => &$value){
        $value['username']  = $key;
      }
      $users  = array_values($users);
      $this->__log(array(__METHOD__, "Found.", $users[0]));
      return $users[0];
    } 
    // if user count more than 1
    if(count($users)>1){
      $this->__log(array(__METHOD__, "$serial for serial found multiple users, that should not happen"));
      return false;
    }
    // no users found
    $this->__log(array(__METHOD__, "$serial not found"));
    return false;    
  }



  /**
  * ____call - delegates all calls to this object to its parents
  *
  * This function needs futher think-through - currently it cant be used so simple, because other classes are also using functionality provided by this - 
  * First none null value is returned - but on all the action is executed, meaning: if you have 4 backends array('plain', 'mysql', 'ad', 'ldap') first values are returned
  * It's more for searching users from multiple backends, not for managing them (but it should manage doing it)
  *
  * @param string $method_name - what was called
  * @param array $method_arguments - what arguments
  * @returns array of results
  */
  function ____call($method_name, $method_arguments){
    global $conf;


    // if $conf['auth']['smartcard']['use_authtypes'] not valid
    if(!isset($conf['auth']['smartcard']['use_authtypes'])){
      // log and return false
      $this->__log(__METHOD__.' \$conf[auth][smartcard][use_authtypes] Must be defined in format: authmodule1,authmodule2,authmodule3.', true, true);
      return false;
    }

    $results  = array();

    // try with different auth modules
    $auth_module_names = preg_split('/,/', $conf['auth']['smartcard']['use_authtypes']);
    foreach($auth_module_names as $auth_module_name){
      // check that name is valid
      if(preg_match('/[^\w\d]/', $auth_module_name) || !$auth_module_name){
        $this->__log(__CLASS__.':'.__METHOD__.' Module name '.$auth_module_name.'is not valid');
        $results[]    = null;
        continue;
      }

      // include class, create object, search for serial and if found return user object
      $include_class_file = DOKU_AUTH.'/'.$auth_module_name.'.class.php';
      $include_class_ok   = include_once($include_class_file);
      $auth_module_class  = 'auth_'.$auth_module_name;

      // check that include went ok
      if(!$include_class_ok){
        // if include went bad log that file was not included, add null value to results and go to new loop
        $this->__log(__CLASS__.':'.__METHOD__.' File not included '.$include_class_file);
        $results[]    = null;
        continue;
      }      
      
      // check that auth class exists
      if(!class_exists($auth_module_class)){
        // if class does not exists log that, add null value to results and go to new loop
        $this->__log(__CLASS__.':'.__METHOD__.' Class not found '.$auth_module_class);
        $results[]    = null;
        continue;
      }

      // check that method exists on the auth module
      if(!method_exists($auth_module_class, $method_name)){
        // add null to results and go to new loop
        $results[]    = null;
        continue;
      }

      $auth_module_ob = new $auth_module_class();      
      $results[]      = call_user_func_array(array($auth_module_ob, $method_name), $method_arguments);
    }    

    $this->__log(array('RESULTS',$method_name,$results));
    return $results;
  }



  /**
  * __call 
  * 
  * for other modules etc first value what evulates to true is returned
  *
  */
  function __call($method_name, $method_arguments=array()){
    $results  = call_user_func_array(array($this, '____call'), array($method_name, $method_arguments));
    foreach($results as $result){
      if($result) return $result;
    }
    // noting did not evulate to true
    return false;
  }



  /**
  * Logs messages to data/log/auth_smartcard.log.txt
  *
  */
  function __log($text, $to_file=true, $to_screen=false){
    global $conf;

    $text = json_encode($text);


    if($to_file && $conf['auth']['smartcard']['log_to_file']){
      @mkdir(DOKU_INC.'/data/pages/log');
      file_put_contents(DOKU_INC.'/data/pages/log/auth_smartcard.log.txt', $text."\n\n", FILE_APPEND);
    }

    if($to_screen){
      msg($text, -1);
    }

  }

}



//Setup VIM: ex: et ts=2 enc=utf-8 :
